packages/server/src/jsRunner.ts

```diff
@@ -1,61 +1,145 @@
+            new ivm.Callback((...params: Parameters<typeof url.resolve>) =>
+              url.resolve(...params)
             )
-          }
-          span?.addTags({
-            js: {
-              limitMS: bbCtx.jsExecutionTracker.limitMs,
-              elapsedMS: bbCtx.jsExecutionTracker.elapsedMS,
-            },
+          )
+
+          global.setSync(
+            "urlParseCb",
+            new ivm.Callback((...params: Parameters<typeof url.parse>) =>
+              url.parse(...params)
+            )
+          )
+
+          const helpersModule = jsIsolate.compileModuleSync(
+            `${injectedRequire};${helpersSource}`
+          )
+
+          const cryptoModule = jsIsolate.compileModuleSync(`export default {
+        randomUUID: cryptoRandomUUIDCb,
+      }`)
+          cryptoModule.instantiateSync(jsContext, specifier => {
+            throw new Error(`No imports allowed. Required: ${specifier}`)
           })
-          // We call checkLimit() here to prevent paying the cost of creating
-          // a new VM context below when we don't need to.
-          tracer.trace("runJS.checkLimitAndBind", {}, span => {
-            bbCtx.jsExecutionTracker!.checkLimit()
-            track = bbCtx.jsExecutionTracker!.track.bind(
-              bbCtx.jsExecutionTracker
+
+          global.setSync(
+            "cryptoRandomUUIDCb",
+            new ivm.Callback(
+              (...params: Parameters<typeof crypto.randomUUID>) => {
+                return crypto.randomUUID(...params)
+              }
             )
+          )
+
+          helpersModule.instantiateSync(jsContext, specifier => {
+            if (specifier === "crypto") {
+              return cryptoModule
+            }
+            throw new Error(`No imports allowed. Required: ${specifier}`)
           })
+
+          for (const [key, value] of Object.entries(ctx)) {
+            if (key === "helpers") {
+              // Can't copy the native helpers into the isolate. We just ignore them as they are handled properly from the helpersSource
+              continue
+            }
+            global.setSync(key, value)
+          }
+
+          bbCtx.isolateRefs = { jsContext, jsIsolate, helpersModule }
         }
-      }
 
-      ctx = {
-        ...ctx,
-        alert: undefined,
-        setInterval: undefined,
-        setTimeout: undefined,
-      }
+        let { jsIsolate, jsContext, helpersModule } = bbCtx.isolateRefs!
+
+        const perRequestLimit = env.JS_PER_REQUEST_TIME_LIMIT_MS
+        if (perRequestLimit) {
+          const cpuMs = Number(jsIsolate.cpuTime) / 1e6
+          if (cpuMs > perRequestLimit) {
+            throw new ExecutionTimeoutError(
+              `CPU time limit exceeded (${cpuMs}ms > ${perRequestLimit}ms)`
+            )
+          }
+        }
+
+        // We need to warp up the actual run in a `cb` function to be able to extract its value from isolated-vm
+        js = js.replace(/run\(\);$/, "cb(run());")
```

Comment:
This makes me a little bit nervous. It would be possible to break JS handling in the server by making a seemingly-unrelated change in the `string-templates` package. Might be worth adding a comment over there saying that the name `run` is important for the backend.

Reply:
We could just tie these two together using a constant somewhere like `const JS_RUNNER_FUNCTION = "run"` then use that in string-templates + here, if someone ever changed it they'd be aware of its impact here as well and ideally it wouldn't be as impacted.

Reply:
I thought the same, and this comment made me rethink it. I used a selfinvoking function in string-templates, that should not break anything and make this usage much easier :)